src/session/async_session.rs

```diff
@@ -327,6 +350,34 @@ impl<S: AsyncRead + Unpin> Stream<S> {
         }
     }
 
+    async fn expect_eager<N: Needle>(&mut self, needle: N) -> Result<Captures, Error> {
+        let start = time::Instant::now();
+        loop {
+            if let Some(timeout) = self.expect_timeout {
+                if start.elapsed() > timeout {
+                    return Err(Error::ExpectTimeout);
+                }
+            }
+
+            let eof = self.stream.read_available().await?;
+            let data = self.stream.get_available();
+
+            let found = expect.check(data, eof)?;
+            if !found.is_empty() {
+                let end_index = Found::right_most_index(&found);
+                let involved_bytes = data[..end_index].to_vec();
+                self.stream.consume_from_buffer(end_index);
+                return Ok(Found::new(involved_bytes, found));
+            } else if !data.is_empty() {
+                let data_len = data.len();
+                self.stream.consume_from_buffer(data_len);
+                continue;
```

Comment:
Is this what you wan't?

I mean potentially a mach will never be found because we drop data here.

Here's an example.

The output `Hello World`.
We do `expect_eager("Hello World")`
First `read_available` reads only `Hello`,
Second `read_available` reads ` World`.
Because you drop the buffer the match will never found even though it was there.

Reply:
I was thinking, maybe we could still drain the unmatched buffer parts up to line breaks that would keep the buffer from growing until the match is found, but only if the Needle itself doesn't contain a line break